modules/cuckooml/cuckooml.py

```diff
@@ -0,0 +1,1677 @@
+        if labels is None:
+            labels = self.labels
+
+        tsne = TSNE(learning_rate=learning_rate)
+        tsne_fit = tsne.fit_transform(data)
+        tsne_df = pd.DataFrame(tsne_fit, index=data.index, columns=['0', '1'])
+        tsne_dfl = pd.concat([tsne_df, labels], axis=1)
+
+        sns.lmplot("0", "1", data=tsne_dfl, fit_reg=False, hue="label",
+                   scatter_kws={"marker":"D", "s":50}, legend_out=True)
+        plt.title(fig_name + " (lr:" + str(learning_rate) + ")")
+        if self.context == "notebook":
+            plt.show()
+        else:
+            plt.savefig(fig_name + "_" + str(learning_rate) + ".png",
+                        bbox_inches='tight', pad_inches=1.)
+            plt.close()
+
+
+    def save_dataset(self, filename="custom_dataset.csv", features=None, \
+                     labels=None):
+        """Export a dataset to CSV formatted file."""
+        # Check if data and labels are loaded
+        if features is None:
+            print "You must indicate data to be saved."
+            return
+
+        if labels is None:
+            print "You didn't indicate labels to be used. Internal labels will \
+                   be used."
+            if self.labels is None:
+                print "Internal labels not available."
+                return
+            else:
+                labels = self.labels
+
+        dataset = pd.concat([features, labels], axis=1)
+        dataset.to_csv(filename, encoding='utf-8')
+
+
+    def cluster_dbscan(self, features=None, eps=20.0, min_samples=5, dry=False):
+        """Do *dbscan* clustering and return """
+        if features is None:
+            print "You didn't indicate features to be used. Internal features \
+                will be used."
+            if self.features is None:
+                print "Internal features not available."
+                return
+            else:
+                features = self.features
+
+        dbscan = DBSCAN(eps=eps, min_samples=min_samples).fit(features)
+
+        # TODO: Allow storing multiple clustering results based on parameters
+        if dry:
+            return {
+                "eps":eps,
+                "min_samples":min_samples,
+                "clustering":pd.DataFrame(dbscan.labels_, index=features.index,
+                                        columns=["label"])
+            }
+        else:
+            # if "dbscan" not in self.clustering:
+                # self.clustering["dbscan"] = {}
+            # clustering_hash = "eps:%f&min_samples:%d" % (eps, min_samples)
+            self.clustering["dbscan"] = {
+            # self.clustering["dbscan"][clustering_hash] = {
+                "eps":eps,
+                "min_samples":min_samples,
+                "clustering":pd.DataFrame(dbscan.labels_, index=features.index,
+                                        columns=["label"])
+            }
+
+
+    def cluster_hdbscan(self, features=None, min_samples=1, \
```

Comment:
@jbremer it largely depends on the number of datapoints to be clustered; but in general (with, say, reasonable number of datapoints) it shouldn't be resource expensive.